src/main/java/dev/openfeature/sdk/OpenFeatureAPI.java

```diff
@@ -83,7 +96,14 @@ public EvaluationContext getEvaluationContext() {
      * Set the default provider.
      */
     public void setProvider(FeatureProvider provider) {
-        providerRepository.setProvider(provider);
+        try (AutoCloseableLock __ = lock.writeLockAutoCloseable()) {
+            providerRepository.setProvider(
+                    provider,
+                    (p) -> attachEventProvider(p),
+                    (p) -> emitReady(p),
+                    (p) -> detachEventProvider(p),
+                    (p, message) -> emitError(p, message));
```

Comment:
for readability, 

```suggestion
                    this::attachEventProvider,
                    this::emitReady,
                    this::detachEventProvider,
                    this::emitError)
```

Reply:
https://github.com/open-feature/java-sdk/pull/476#discussion_r1244194878